apps/heft/src/stages/BuildStage.ts

```diff
@@ -0,0 +1,407 @@
+      locale: standardParameters.localeParameter.value,
+      typescriptMaxWriteParallelism: standardParameters.typescriptMaxWriteParallelismParamter.value
+    };
+  }
+
+  protected getDefaultStageProperties(options: IBuildStageOptions): IBuildStageProperties {
+    return {
+      production: options.production,
+      lite: options.lite,
+      locale: options.locale,
+      maxOldSpaceSize: options.maxOldSpaceSize,
+      watchMode: options.watchMode,
+      serveMode: options.serveMode
+    };
+  }
+
+  protected async executeInnerAsync(): Promise<void> {
+    const preCompileSubstage: IPreCompileSubstage = {
+      hooks: new BuildSubstageHooksBase(),
+      properties: {}
+    };
+    this.stageHooks.preCompile.call(preCompileSubstage);
+
+    const compileStage: ICompileSubstage = {
+      hooks: new CompileSubstageHooks(),
+      properties: {
+        typeScriptConfiguration: {
+          tsconfigPaths: [],
+          isLintingEnabled: !this.stageProperties.lite,
+          copyFromCacheMode: undefined,
+          additionalModuleKindsToEmit: undefined,
+          maxWriteParallelism: 50
+        },
+        copyStaticAssetsConfiguration: {
+          fileExtensions: [],
+          excludeGlobs: [],
+          includeGlobs: [],
+
+          // For now - these may need to be revised later
+          sourceFolderName: 'src',
+          destinationFolderNames: ['lib']
+        }
+      }
+    };
+    this.stageHooks.compile.call(compileStage);
+
+    const bundleStage: IBundleSubstage = {
+      hooks: new BundleSubstageHooks(),
+      properties: {}
+    };
+    this.stageHooks.bundle.call(bundleStage);
+
+    const postBuildStage: IPostBuildSubstage = {
+      hooks: new BuildSubstageHooksBase(),
+      properties: {}
+    };
+    this.stageHooks.postBuild.call(postBuildStage);
+
+    if (this.stageProperties.watchMode) {
+      // In --watch mode, run all configuration upfront and then kick off all stages
+      // concurrently with the expectation that the their promises will never resolve
+      // and that they will handle watching filesystem changes
+
+      await Promise.all([
+        compileStage.hooks.configureTypeScript.promise(),
+        compileStage.hooks.configureCopyStaticAssets.promise(),
+        bundleStage.hooks.configureWebpack.promise()
+      ]);
+      await Promise.all([
+        compileStage.hooks.afterConfigureTypeScript.promise(),
+        compileStage.hooks.afterConfigureCopyStaticAssets.promise(),
+        bundleStage.hooks.afterConfigureWebpack.promise()
+      ]);
+
+      await Promise.all([
+        this._runSubstageWithLoggingAsync('Pre-compile', preCompileSubstage),
+        this._runSubstageWithLoggingAsync('Compile', compileStage),
+        this._runSubstageWithLoggingAsync('Bundle', bundleStage),
+        this._runSubstageWithLoggingAsync('Post-build', postBuildStage)
+      ]);
+    } else {
+      await this._runSubstageWithLoggingAsync('Pre-compile', preCompileSubstage);
```

Comment:
👍 👍  Separating out the stages into actual source code objects is a huge improvement.